File: packages/go/cbor.go
package flatwire

// Streaming CBOR (RFC 8949 binary) format for flatwire, mirroring the Python
// reference. flatwire's CBOR wire is a stream of concatenated CBOR data items
// (not a length-prefixed array), so encoding needs no upfront count and decoding
// reads one item at a time. The encoding is deterministic (shortest heads, map
// keys sorted by UTF-8 bytes, 64-bit floats), so output is byte-identical across
// all six flatwire languages. Covers the JSON data model
// (null/bool/int/float/str/array/map); no tags.
//
// Values are Go's generic any: objects -> map[string]any, arrays -> []any,
// scalars -> int64/uint64/float64/bool/string/nil.

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"sort"
)

// --- encoding --------------------------------------------------------------

func writeCborHead(w io.Writer, major byte, n uint64) error {
	mt := major << 5
	switch {
	case n < 24:
		_, err := w.Write([]byte{mt | byte(n)})
		return err
	case n <= 0xff:
		_, err := w.Write([]byte{mt | 24, byte(n)})
		return err
	case n <= 0xffff:
		var b [3]byte
		b[0] = mt | 25
		binary.BigEndian.PutUint16(b[1:], uint16(n))
		_, err := w.Write(b[:])
		return err
	case n <= 0xffffffff:
		var b [5]byte
		b[0] = mt | 26
		binary.BigEndian.PutUint32(b[1:], uint32(n))
		_, err := w.Write(b[:])
		return err
	default:
		var b [9]byte
		b[0] = mt | 27
		binary.BigEndian.PutUint64(b[1:], n)
		_, err := w.Write(b[:])
		return err
	}
}

func writeCborValue(w io.Writer, v any) error {
	switch x := v.(type) {
	case nil:
		_, err := w.Write([]byte{0xf6})
		return err
	case bool:
		if x {
			_, err := w.Write([]byte{0xf5})
			return err
		}
		_, err := w.Write([]byte{0xf4})
		return err
	case string:
		if err := writeCborHead(w, 3, uint64(len(x))); err != nil {
			return err
		}
		_, err := io.WriteString(w, x)
		return err
	case float32:
		return writeCborF64(w, float64(x))
	case float64:
		return writeCborF64(w, x)
	case int:
		return writeCborInt(w, int64(x))
	case int8:
		return writeCborInt(w, int64(x))
	case int16:
		return writeCborInt(w, int64(x))
	case int32:
		return writeCborInt(w, int64(x))
	case int64:
		return writeCborInt(w, x)
	case uint:
		return writeCborHead(w, 0, uint64(x))
	case uint8:
		return writeCborHead(w, 0, uint64(x))
	case uint16:
		return writeCborHead(w, 0, uint64(x))
	case uint32:
		return writeCborHead(w, 0, uint64(x))
	case uint64:
		return writeCborHead(w, 0, x)
	case []any:
		if err := writeCborHead(w, 4, uint64(len(x))); err != nil {
			return err
		}
		for _, e := range x {
			if err := writeCborValue(w, e); err != nil {
				return err
			}
		}
		return nil
	case map[string]any:
		if err := writeCborHead(w, 5, uint64(len(x))); err != nil {
			return err
		}
		// Deterministic: sort keys by their UTF-8 byte sequence (Go string
		// ordering is byte-wise on UTF-8).
		keys := make([]string, 0, len(x))
		for k := range x {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			if err := writeCborHead(w, 3, uint64(len(k))); err != nil {
				return err
			}
			if _, err := io.WriteString(w, k); err != nil {
				return err
			}
			if err := writeCborValue(w, x[k]); err != nil {
				return err
			}
		}
		return nil
	default:
		return fmt.Errorf("flatwire cbor: unsupported type %T", v)
	}
}

func writeCborInt(w io.Writer, v int64) error {
	if v >= 0 {
		return writeCborHead(w, 0, uint64(v))
	}
	return writeCborHead(w, 1, uint64(-1-v))
}

func writeCborF64(w io.Writer, f float64) error {
	var b [9]byte
	b[0] = 0xfb
	binary.BigEndian.PutUint64(b[1:], math.Float64bits(f))
	_, err := w.Write(b[:])
	return err
}

// EncodeArrayCBOR streams a collection as concatenated CBOR data items, one per
// element. Peak memory is bounded by the largest single element. Returns the
// element count.
func EncodeArrayCBOR(items []any, w io.Writer) (int, error) {
	count := 0
	for _, v := range items {
		if err := writeCborValue(w, v); err != nil {
			return count, err
		}
		count++
	}
	return count, nil
}

// --- decoding --------------------------------------------------------------

// DecodeArrayCBOR lazily reads concatenated CBOR data items, calling yield with
// each decoded element in turn. Backed by a buffered reader, so the whole stream
// is never held in memory at once.
func DecodeArrayCBOR(r io.Reader, yield func(any) error) error {
	br := bufio.NewReaderSize(r, 65536)
	for {
		_, err := br.Peek(1)
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		v, err := readCborValue(br)
		if err != nil {
			return err
		}
		if err := yield(v); err != nil {
			return err
		}
	}
}

func readCborArgument(br *bufio.Reader, ai byte) (uint64, error) {
	switch {
	case ai < 24:
		return uint64(ai), nil
	case ai == 24:
		b, err := br.ReadByte()
		return uint64(b), err
	case ai == 25:
		b, err := readN(br, 2)
		if err != nil {
			return 0, err
		}
		return uint64(binary.BigEndian.Uint16(b)), nil
	case ai == 26:
		b, err := readN(br, 4)
		if err != nil {
			return 0, err
		}
		return uint64(binary.BigEndian.Uint32(b)), nil
	case ai == 27:
		b, err := readN(br, 8)
		if err != nil {
			return 0, err
		}
		return binary.BigEndian.Uint64(b), nil
	default:
		return 0, fmt.Errorf("flatwire cbor: unsupported additional info %d", ai)
	}
}

func readCborValue(br *bufio.Reader) (any, error) {
	ib, err := br.ReadByte()
	if err != nil {
		return nil, err
	}
	major := ib >> 5
	ai := ib & 0x1f
	switch major {
	case 0:
		n, err := readCborArgument(br, ai)
		if err != nil {
			return nil, err
		}
		if n <= math.MaxInt64 {
			return int64(n), nil
		}
		return n, nil
	case 1:
		n, err := readCborArgument(br, ai)
		if err != nil {
			return nil, err
		}
		return int64(-1 - int64(n)), nil
	case 2:
		return nil, fmt.Errorf("flatwire cbor: byte-string is not part of the JSON value model")
	case 3:
		n, err := readCborArgument(br, ai)
		if err != nil {
			return nil, err
		}
		b, err := readN(br, int(n))
		if err != nil {
			return nil, err
		}
		return string(b), nil
	case 4:
		n, err := readCborArgument(br, ai)
		if err != nil {
			return nil, err
		}
		arr := make([]any, n)
		for i := 0; i < int(n); i++ {
			e, err := readCborValue(br)
			if err != nil {
				return nil, err
			}
			arr[i] = e
		}
		return arr, nil
	case 5:
		n, err := readCborArgument(br, ai)
		if err != nil {
			return nil, err
		}
		m := make(map[string]any, n)
		for i := 0; i < int(n); i++ {
			k, err := readCborValue(br)
			if err != nil {
				return nil, err
			}
			val, err := readCborValue(br)
			if err != nil {
				return nil, err
			}
			key, ok := k.(string)
			if !ok {
				key = fmt.Sprintf("%v", k)
			}
			m[key] = val
		}
		return m, nil
	case 7:
		switch ai {
		case 20:
			return false, nil
		case 21:
			return true, nil
		case 22:
			return nil, nil
		case 23:
			return nil, nil // undefined -> null
		case 25:
			b, err := readN(br, 2)
			if err != nil {
				return nil, err
			}
			return decodeCborF16(binary.BigEndian.Uint16(b)), nil
		case 26:
			b, err := readN(br, 4)
			if err != nil {
				return nil, err
			}
			return float64(math.Float32frombits(binary.BigEndian.Uint32(b))), nil
		case 27:
			b, err := readN(br, 8)
			if err != nil {
				return nil, err
			}
			return math.Float64frombits(binary.BigEndian.Uint64(b)), nil
		default:
			return nil, fmt.Errorf("flatwire cbor: unsupported simple value %d", ai)
		}
	default:
		return nil, fmt.Errorf("flatwire cbor: unsupported major type %d", major)
	}
}

func decodeCborF16(h uint16) float64 {
	sign := (h >> 15) & 0x1
	exp := (h >> 10) & 0x1f
	frac := h & 0x3ff
	var val float64
	switch {
	case exp == 0:
		val = (float64(frac) / 1024.0) * math.Pow(2, -14)
	case exp == 0x1f:
		if frac == 0 {
			val = math.Inf(1)
		} else {
			val = math.NaN()
		}
	default:
		val = (1.0 + float64(frac)/1024.0) * math.Pow(2, float64(exp)-15)
	}
	if sign == 1 {
		return -val
	}
	return val
}

File: packages/python/flatwire/cbor.py
"""Streaming CBOR (RFC 8949 binary) format for flatwire.

Like the MessagePack format, a CBOR *array* is length-prefixed, which would force
us to know the element count before writing anything - incompatible with
streaming an open-ended iterable. So flatwire's CBOR wire is a **stream of
concatenated CBOR data items**: each element is encoded as one self-describing
CBOR value, written back-to-back. The decoder reads exactly one item at a time
until the stream ends. This keeps encode memory flat (no buffering to count
elements) and shrinks the bytes on the wire versus JSON.

The encoding is **deterministic** so the output is byte-identical across all six
flatwire languages (RFC 8949 4.2 core deterministic rules, applied to the JSON
data model):

- integers use the shortest head that fits;
- text/byte-string and array/map heads use the shortest length encoding;
- map keys are sorted by their UTF-8 byte sequence;
- floats are always encoded as 64-bit (`0xFB`) - simple, and still round-trips.

This is a focused, dependency-free codec covering the JSON data model
(null/bool/int/float/str/bytes/array/map). It is spec-correct for those types; it
is not a full CBOR implementation (no tags, no indefinite-length items on encode).
"""

from __future__ import annotations

import struct
from typing import Any, BinaryIO, Iterable, Iterator


# --- encoding --------------------------------------------------------------

def _head(major: int, n: int, out: bytearray) -> None:
    """Write a CBOR head: the major type (0-7) with the shortest argument
    encoding for ``n`` (RFC 8949 deterministic rule)."""
    mt = major << 5
    if n < 24:
        out.append(mt | n)
    elif n <= 0xFF:
        out.append(mt | 24)
        out.append(n)
    elif n <= 0xFFFF:
        out.append(mt | 25)
        out.extend(struct.pack(">H", n))
    elif n <= 0xFFFFFFFF:
        out.append(mt | 26)
        out.extend(struct.pack(">I", n))
    elif n <= 0xFFFFFFFFFFFFFFFF:
        out.append(mt | 27)
        out.extend(struct.pack(">Q", n))
    else:
        raise OverflowError("flatwire cbor: value out of 64-bit range")


def _encode_value(v: Any, out: bytearray) -> None:
    if v is None:
        out.append(0xF6)
    elif v is True:
        out.append(0xF5)
    elif v is False:
        out.append(0xF4)
    elif isinstance(v, int):
        # major 0 for non-negative, major 1 for negative (argument = -1 - v).
        if v >= 0:
            _head(0, v, out)
        else:
            _head(1, -1 - v, out)
    elif isinstance(v, float):
        out.append(0xFB)
        out.extend(struct.pack(">d", v))
    elif isinstance(v, str):
        b = v.encode("utf-8")
        _head(3, len(b), out)
        out.extend(b)
    elif isinstance(v, (bytes, bytearray)):
        b = bytes(v)
        _head(2, len(b), out)
        out.extend(b)
    elif isinstance(v, dict):
        _encode_map(v, out)
    elif isinstance(v, (list, tuple)):
        _head(4, len(v), out)
        for item in v:
            _encode_value(item, out)
    else:
        raise TypeError(f"flatwire cbor: unsupported type {type(v).__name__}")


def _encode_map(v: dict, out: bytearray) -> None:
    # Deterministic: sort entries by the key's UTF-8 byte sequence so the
    # encoding is byte-identical regardless of the source map's iteration order.
    items = sorted(v.items(), key=lambda kv: str(kv[0]).encode("utf-8"))
    _head(5, len(items), out)
    for k, val in items:
        _encode_value(k, out)
        _encode_value(val, out)


def encode_array(items: Iterable[Any], fp: BinaryIO) -> int:
    """Stream a collection as concatenated CBOR data items, one per element.

    Peak memory is bounded by the largest single element. Returns the count.
    """
    count = 0
    for item in items:
        buf = bytearray()
        _encode_value(item, buf)
        fp.write(buf)
        count += 1
    return count


# --- decoding --------------------------------------------------------------

class _Reader:
    """Buffers the stream and reads CBOR values, refilling as needed so a value
    split across reads is handled. Consumed bytes are dropped so memory stays
    bounded by the largest single element."""

    def __init__(self, fp: BinaryIO, chunk_size: int = 65536):
        self.fp = fp
        self.buf = bytearray()
        self.pos = 0
        self.chunk = chunk_size
        self.eof = False

    def _fill(self, need: int) -> bool:
        while len(self.buf) - self.pos < need:
            data = self.fp.read(self.chunk)
            if not data:
                self.eof = True
                return len(self.buf) - self.pos >= need
            self.buf.extend(data)
        return True

    def at_end(self) -> bool:
        if self.pos < len(self.buf):
            return False
        data = self.fp.read(self.chunk)
        if not data:
            self.eof = True
            return True
        self.buf.extend(data)
        return False

    def _take(self, n: int) -> bytes:
        if not self._fill(n):
            raise EOFError("flatwire cbor: truncated value")
        b = bytes(self.buf[self.pos:self.pos + n])
        self.pos += n
        if self.pos > self.chunk:
            del self.buf[:self.pos]
            self.pos = 0
        return b

    def u8(self) -> int:
        return self._take(1)[0]

    def _argument(self, ai: int) -> int:
        if ai < 24:
            return ai
        if ai == 24:
            return self.u8()
        if ai == 25:
            return struct.unpack(">H", self._take(2))[0]
        if ai == 26:
            return struct.unpack(">I", self._take(4))[0]
        if ai == 27:
            return struct.unpack(">Q", self._take(8))[0]
        raise ValueError(f"flatwire cbor: unsupported additional info {ai}")

    def read_value(self) -> Any:
        ib = self.u8()
        major = ib >> 5
        ai = ib & 0x1F
        if major == 0:  # unsigned int
            return self._argument(ai)
        if major == 1:  # negative int
            return -1 - self._argument(ai)
        if major == 2:  # byte string
            return self._take(self._argument(ai))
        if major == 3:  # text string
            return self._take(self._argument(ai)).decode("utf-8")
        if major == 4:  # array
            n = self._argument(ai)
            return [self.read_value() for _ in range(n)]
        if major == 5:  # map
            n = self._argument(ai)
            out = {}
            for _ in range(n):
                k = self.read_value()
                out[k] = self.read_value()
            return out
        if major == 7:  # simple / float
            if ai == 20:
                return False
            if ai == 21:
                return True
            if ai == 22:
                return None
            if ai == 23:
                return None  # undefined -> null in the JSON data model
            if ai == 25:
                return _decode_float16(self._take(2))
            if ai == 26:
                return struct.unpack(">f", self._take(4))[0]
            if ai == 27:
                return struct.unpack(">d", self._take(8))[0]
            raise ValueError(f"flatwire cbor: unsupported simple value {ai}")
        raise ValueError(f"flatwire cbor: unsupported major type {major}")


def _decode_float16(b: bytes) -> float:
    (h,) = struct.unpack(">H", b)
    sign = (h >> 15) & 0x1
    exp = (h >> 10) & 0x1F
    frac = h & 0x3FF
    if exp == 0:
        val = (frac / 1024.0) * (2.0 ** -14)
    elif exp == 0x1F:
        val = float("inf") if frac == 0 else float("nan")
    else:
        val = (1.0 + frac / 1024.0) * (2.0 ** (exp - 15))
    return -val if sign else val


def decode_array(fp: BinaryIO, chunk_size: int = 65536) -> Iterator[Any]:
    """Lazily read concatenated CBOR data items, yielding one at a time.

    Reads exactly one value per iteration and drops consumed bytes, so peak
    memory stays proportional to the largest single element.
    """
    r = _Reader(fp, chunk_size=chunk_size)
    while not r.at_end():
        yield r.read_value()
